File: upande_sphynx/upande_sphynx/report/accounts_payable_aging/accounts_payable_aging.js
// Copyright (c) 2026, Jeniffer and contributors
// For license information, please see license.tx

frappe.query_reports["Accounts Payable Aging"] = {

	// ── Filters ───────────────────────────────────────────────────────────────
	filters: [
		{
			fieldname: "company",
			label:     __("Company"),
			fieldtype: "Link",
			options:   "Company",
			reqd:      1,
			default:   frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "from_date",
			label:     __("From Date"),
			fieldtype: "Date",
			reqd:      1,
			default:   frappe.datetime.year_start(),   // 1 Jan of current year
		},
		{
			fieldname: "to_date",
			label:     __("To Date"),
			fieldtype: "Date",
			reqd:      1,
			default:   frappe.datetime.get_today(),
		},
		{
			fieldname: "party_account",
			label:     __("Payable Account"),
			fieldtype: "Link",
			options:   "Account",
			get_query: () => {
				const company = frappe.query_report.get_filter_value("company");
				return {
					filters: {
						company:      company,
						account_type: "Payable",
						is_group:     0,
					},
				};
			},
		},
		{
			fieldname: "party",
			label:     __("Supplier(s)"),
			fieldtype: "MultiSelectList",
			get_data:  function (txt) {
				return frappe.db.get_link_options("Supplier", txt);
			},
		},
		{
			fieldname: "supplier_group",
			label:     __("Supplier Group"),
			fieldtype: "Link",
			options:   "Supplier Group",
		},
	],

	// ── Row formatter ─────────────────────────────────────────────────────────
	formatter: function (value, row, column, data, default_formatter) {
		if (!data) return default_formatter(value, row, column, data);

		// Currency columns — show symbol + 2dp, never a raw float
		const currency_cols = [
			"grand_total", "paid_amount", "outstanding_amount",
			"range1", "range2", "range3", "range4", "range5",
		];
		if (currency_cols.includes(column.fieldname)) {
			const num = flt(value);
			if (num === 0) return `<span class="ap-zero">—</span>`;
			const symbol = (data.currency && frappe.get_currency_symbol)
				? frappe.get_currency_symbol(data.currency)
				: (data.currency || "");
			const formatted = num.toLocaleString("en-US", {
				minimumFractionDigits: 2,
				maximumFractionDigits: 2,
			});
			return `<span class="ap-num">${symbol} ${formatted}</span>`;
		}

		// Status badge
		if (column.fieldname === "status") {
			if (value === "Overdue") {
				return `<span class="ap-badge ap-overdue">Overdue</span>`;
			}
			return `<span class="ap-badge ap-current">Current</span>`;
		}

		// Aging bucket pill
		if (column.fieldname === "aging_bucket") {
			const cls = {
				"0-30":   "ap-b0",
				"31-60":  "ap-b1",
				"61-90":  "ap-b2",
				"91-120": "ap-b3",
				"121+":   "ap-b4",
			}[value] || "ap-b0";
			return `<span class="ap-bucket ${cls}">${value}</span>`;
		}

		// Bold totals
		if (data.bold) {
			return `<strong>${default_formatter(value, row, column, data)}</strong>`;
		}

		return default_formatter(value, row, column, data);
	},

	// ── Chart ─────────────────────────────────────────────────────────────────
	get_chart_data: function (columns, result) {
		if (!result || !result.length) return;

		const fields  = ["range1", "range2", "range3", "range4", "range5"];
		const labels  = ["0-30", "31-60", "61-90", "91-120", "121+"];
		const totals  = [0, 0, 0, 0, 0];
		let currency  = "";

		result.forEach((row) => {
			if (!row.invoice_no) return;              // skip summary rows
			if (!currency && row.currency) currency = row.currency;
			fields.forEach((f, i) => { totals[i] += flt(row[f]); });
		});

		const symbol = (currency && frappe.get_currency_symbol)
			? frappe.get_currency_symbol(currency)
			: currency;

		return {
			data: {
				labels:   labels,
				datasets: [{ name: __("Outstanding"), values: totals }],
			},
			type:   "bar",
			colors: ["#e05c5c"],
			height: 240,
			axisOptions: { xIsSeries: true },
			tooltipOptions: {
				formatTooltipY: (v) =>
					symbol + " " + Number(v).toLocaleString("en-US", {
						minimumFractionDigits: 2, maximumFractionDigits: 2,
					}),
			},
			title: __("Outstanding by Aging Bucket") + (currency ? ` (${currency})` : ""),
		};
	},

	// ── Toolbar buttons ───────────────────────────────────────────────────────
	onload: function (report) {
		inject_ap_styles();

		// AP Summary button — clicking it passes current filters and adds a
		// back-link so AP Summary's "Accounts Payable" button comes back here.
		report.page.add_inner_button(__("AP Summary"), function () {
			const f = report.get_values();
			frappe.set_route("query-report", "Accounts Payable Summary", {
				company: f.company,
			});
		});

		// Override the AP Summary report's back-button label so it points
		// to this custom report instead of the default Accounts Payable.
		// We do this by monkey-patching after the route change settles.
		frappe.router.on("change", function () {
			const route = frappe.get_route();
			if (
				route[0] === "query-report" &&
				route[1] === "Accounts Payable Summary"
			) {
				setTimeout(() => {
					// Replace the built-in "Accounts Payable" button with one
					// that routes back to our custom aging report.
					const $btn = $(".page-head .inner-toolbar button")
						.filter((_, el) =>
							$(el).text().trim() === "Accounts Payable"
						);
					if ($btn.length) {
						$btn.text(__("AP Aging (Custom)")).off("click").on("click", () => {
							frappe.set_route("query-report", "Accounts Payable Aging");
						});
					}
				}, 800);
			}
		});
	},

	// Inject styles after datatable renders
	after_datatable_render: function () {
		inject_ap_styles();
	},
};

// ── Style injection (runs once) ───────────────────────────────────────────────
function inject_ap_styles() {
	if (document.getElementById("ap-aging-styles")) return;
	const s = document.createElement("style");
	s.id = "ap-aging-styles";
	s.textContent = `
		/* currency */
		.ap-num  { display:block; text-align:right; font-variant-numeric:tabular-nums; }
		.ap-zero { display:block; text-align:right; color:#bbb; }

		/* status badges */
		.ap-badge {
			display:inline-block; padding:2px 9px; border-radius:20px;
			font-size:11px; font-weight:600; letter-spacing:.04em; text-transform:uppercase;
		}
		.ap-overdue { background:#fde8e8; color:#a02020; border:1px solid #f5b7b1; }
		.ap-current { background:#eafaf1; color:#1e6b3a; border:1px solid #a9dfbf; }

		/* aging bucket pills */
		.ap-bucket {
			display:inline-block; padding:2px 8px; border-radius:4px;
			font-size:11px; font-weight:600; font-family:monospace;
		}
		.ap-b0 { background:#eafaf1; color:#1e6b3a; }
		.ap-b1 { background:#fef9e7; color:#8a6d00; }
		.ap-b2 { background:#fff3e8; color:#a04000; }
		.ap-b3 { background:#fde8e8; color:#a02020; }
		.ap-b4 { background:#f2ecf8; color:#5a2d82; }
	`;
	document.head.appendChild(s);
}

File: upande_sphynx/upande_sphynx/report/accounts_payable_aging/accounts_payable_aging.py
# Copyright (c) 2026, Jeniffer and contributors
# For license information, please see license.txt



import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate, date_diff


def execute(filters=None):
    filters = frappe._dict(filters or {})

    if not filters.from_date:
        filters.from_date = frappe.defaults.get_user_default("year_start_date") or "2000-01-01"
    if not filters.to_date:
        filters.to_date = nowdate()

    if getdate(filters.from_date) > getdate(filters.to_date):
        frappe.throw(_("From Date cannot be after To Date"))

    columns = get_columns()
    data    = get_data(filters)
    return columns, data


# ── Columns ───────────────────────────────────────────────────────────────────
def get_columns():
    return [
        {
            "label":     _("Supplier"),
            "fieldname": "supplier",
            "fieldtype": "Link",
            "options":   "Supplier",
            "width":     190,
        },
		{
            "label":     _("Posting Date"),
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width":     120,
        },
        {
            "label":     _("Due Date"),
            "fieldname": "due_date",
            "fieldtype": "Date",
            "width":     120,
        },
        {
            "label":     _("Invoice No"),
            "fieldname": "invoice_no",
            "fieldtype": "Link",
            "options":   "Purchase Invoice",
            "width":     160,
        },
		
		{
            "label":     _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width":     95,
        },

        {
            "label":     _("Outstanding"),
            "fieldname": "outstanding_amount",
            "fieldtype": "Currency",
            "options":   "currency",
            "width":     170,
        },
        
        {
            "label":     _("Age (Days)"),
            "fieldname": "age_days",
            "fieldtype": "Int",
            "width":     90,
        },
        # {
        #     "label":     _("Aging Bucket"),
        #     "fieldname": "aging_bucket",
        #     "fieldtype": "Data",
        #     "width":     105,
        # },
        {
            "label":     _("0-30"),
            "fieldname": "range1",
            "fieldtype": "Currency",
            "options":   "currency",
            "width":     120,
        },
        {
            "label":     _("31-60"),
            "fieldname": "range2",
            "fieldtype": "Currency",
            "options":   "currency",
            "width":     120,
        },
        {
            "label":     _("61-90"),
            "fieldname": "range3",
            "fieldtype": "Currency",
            "options":   "currency",
            "width":     120,
        },
        {
            "label":     _("91-120"),
            "fieldname": "range4",
            "fieldtype": "Currency",
            "options":   "currency",
            "width":     120,
        },
        {
            "label":     _("121+"),
            "fieldname": "range5",
            "fieldtype": "Currency",
            "options":   "currency",
            "width":     120,
        },
    ]


# ── Data ──────────────────────────────────────────────────────────────────────
def get_data(filters):
    from_date    = getdate(filters.from_date)
    to_date      = getdate(filters.to_date)
    age_ref_date = to_date

    # ── Step 1: fetch all submitted Purchase Invoices in the date range ───────
    inv_conditions = [
        "pi.docstatus = 1",
        "pi.posting_date BETWEEN %(from_date)s AND %(to_date)s",
    ]
    inv_values = {
        "from_date": from_date,
        "to_date":   to_date,
    }

    if filters.get("company"):
        inv_conditions.append("pi.company = %(company)s")
        inv_values["company"] = filters.company

    if filters.get("party_account"):
        inv_conditions.append("pi.credit_to = %(party_account)s")
        inv_values["party_account"] = filters.party_account

    if filters.get("supplier_group"):
        inv_conditions.append("sup.supplier_group = %(supplier_group)s")
        inv_values["supplier_group"] = filters.supplier_group

    party = filters.get("party")
    if isinstance(party, list):
        party = [p for p in party if p]
    if party:
        if isinstance(party, list) and len(party) == 1:
            inv_conditions.append("pi.supplier = %(supplier_single)s")
            inv_values["supplier_single"] = party[0]
        elif isinstance(party, list):
            inv_conditions.append("pi.supplier IN %(party)s")
            inv_values["party"] = tuple(party)
        else:
            inv_conditions.append("pi.supplier = %(party)s")
            inv_values["party"] = party

    invoices = frappe.db.sql(
        """
        SELECT
            pi.name          AS invoice_no,
            pi.supplier,
            pi.posting_date,
            pi.due_date,
            pi.currency,
            pi.grand_total
        FROM
            `tabPurchase Invoice` pi
            LEFT JOIN `tabSupplier` sup ON sup.name = pi.supplier
        WHERE
            {where}
        ORDER BY
            pi.supplier, pi.posting_date
        """.format(where=" AND ".join(inv_conditions)),
        inv_values,
        as_dict=True,
    )

    if not invoices:
        return []

    voucher_nos = [inv.invoice_no for inv in invoices]

    # ── Step 2: reconstruct outstanding as of to_date from GL Entry ───────────
    # For each voucher, sum credits (= invoice postings) and debits
    # (= payments/credit notes) that occurred on or before to_date.
    # outstanding_as_of_to_date = credit_total - debit_total
    gl_rows = frappe.db.sql(
        """
        SELECT
            gle.voucher_no,
            SUM(CASE WHEN gle.credit_in_account_currency > 0
                THEN gle.credit_in_account_currency ELSE 0 END) AS total_credit,
            SUM(CASE WHEN gle.debit_in_account_currency  > 0
                THEN gle.debit_in_account_currency  ELSE 0 END) AS total_debit
        FROM `tabGL Entry` gle
        INNER JOIN `tabAccount` acc ON acc.name = gle.account
        WHERE
            gle.voucher_no   IN %(voucher_nos)s
            AND gle.party_type = 'Supplier'
            AND gle.is_cancelled = 0
            AND acc.account_type = 'Payable'
            AND gle.posting_date <= %(to_date)s
        GROUP BY gle.voucher_no
        """,
        {
            "voucher_nos": tuple(voucher_nos),
            "to_date":     to_date,
        },
        as_dict=True,
    )

    # Build a lookup: voucher_no -> outstanding as of to_date
    gl_map = {}
    for row in gl_rows:
        credit = flt(row.total_credit)
        debit  = flt(row.total_debit)
        gl_map[row.voucher_no] = {
            "outstanding": credit - debit,
            "paid":        debit,
        }

    # ── Step 3: build report rows ─────────────────────────────────────────────
    data = []
    for inv in invoices:
        gl = gl_map.get(inv.invoice_no, {})
        out_amt  = flt(gl.get("outstanding", inv.grand_total))
        paid_amt = flt(gl.get("paid", 0))

        # Skip invoices that were fully settled on or before to_date
        if out_amt <= 0.005:
            continue

        due_date = getdate(inv.due_date or inv.posting_date)
        age_days = date_diff(age_ref_date, due_date)
        status   = "Overdue" if age_days > 0 else "Current"

        if age_days <= 30:
            bucket = "0-30"
        elif age_days <= 60:
            bucket = "31-60"
        elif age_days <= 90:
            bucket = "61-90"
        elif age_days <= 120:
            bucket = "91-120"
        else:
            bucket = "121+"

        data.append({
            "supplier":           inv.supplier,
            "invoice_no":         inv.invoice_no,
            "posting_date":       inv.posting_date,
            "due_date":           due_date,
            "currency":           inv.currency,
            "grand_total":        flt(inv.grand_total),
            "paid_amount":        paid_amt,
            "outstanding_amount": out_amt,
            "status":             status,
            "age_days":           max(age_days, 0),
            "aging_bucket":       bucket,
            "range1":             out_amt if age_days <= 30        else 0,
            "range2":             out_amt if 30  < age_days <= 60  else 0,
            "range3":             out_amt if 60  < age_days <= 90  else 0,
            "range4":             out_amt if 90  < age_days <= 120 else 0,
            "range5":             out_amt if age_days > 120        else 0,
        })

    return data
